README.md
<p align="center">
<img src="https://github.com/Sbozzolo/motionpicture/raw/master/logo.png" width="636" height="131">
</p>

[![GPLv3
license](https://img.shields.io/badge/License-GPLv3-blue.svg)](http://perso.crans.org/besson/LICENSE.html)
![Tests and documentation](https://github.com/Sbozzolo/motionpicture/workflows/Tests/badge.svg)
[![codecov](https://codecov.io/gh/Sbozzolo/motionpicture/branch/master/graph/badge.svg?token=z7jvNNdwVS)](https://codecov.io/gh/Sbozzolo/motionpicture)
[![PyPI version](https://badge.fury.io/py/motionpicture.svg)](https://badge.fury.io/py/motionpicture)

# Introduction

`motionpicture` is a Python library to simplify the creation of videos out of
individual frames. With `motionpicture`, you just have to specify how to produce
a generic frame, and the package will do everything else for you. In
`motionpicture`, your code can be configured via command-line or text files:
turning your code into a plug-in for `motionpicture` is trivial, so you will be
able to reuse your code with ease.

# Examples

There are two important ingredients to use `motionpicture`: `mopi`, and a
_movie_ file. `mopi` is a command-line utility that comes when you install this
package. It will be your main interface to `motionpicture` and it has a
comprehensive `--help` function. A movie file is a recipe on how to produce a
generic frame. With few small restrictions, you have full control over this file
(more info in section [Movie files](#movie-files)).

In these examples we are going to use `matplotlib` to do the plotting, but you
are completely free to generate frames with any Python package you wish.

## Unveiling a sine wave

In this example, we show how to use `mopi` to generate the following video.
![sine_wave](https://github.com/Sbozzolo/motionpicture/raw/master/sine_wave.gif)

To produce this video, we need the following movie file.
``` python
import matplotlib.pyplot as plt
import numpy as np

class MOPIMovie:
    def __init__(self, _args):
        self.times = np.linspace(0, 10, 100)
        self.values = np.sin(self.times)

    def get_frames(self):
        # Here we tell motionpicture what we consider a frame
        return range(self.times)

    def make_frame(self, path, frame_number):
        # Here we plot a specific frame
        plt.clf()
        plt.plot(self.times[:frame_number], self.values[:frame_number])
        plt.xlim([0, self.times[-1]])
        plt.ylim([-1, 1])
        plt.savefig(path)
```
Assuming this file is saved in `sin_wave.py`, we run

``` sh
mopi -m sin_wave.py -o frames_dir --parallel
```
This is produce the individual frames in a folder `frames_dir` using all the CPUs
available on your machine. Then, it will glue the frames together in a video
that has the default name of `video.mp4`. If you want to change name, or other
properties (e.g., the fps), you can add options to `mopi`
``` sh
mopi -m sin_wave.py -o frames_dir --parallel --fps 10 --movie-name sin_wave
```
This will produce a `sin_wave.mp4` video with 10 frames per second instead

## Unveiling a sine wave with controllable frequency

Let us continue on the example of the sine wave, and let us assume that we want
to explore different frequencies.

We can edit the previous movie file adding a `mopi_add_custom_options`
function:

``` python
def mopi_add_custom_options(parser):
    """Add command-line options specific to this movie."""
    parser.add_argument(
        "-f",
        "--frequency",
        default=1,
        type=int,
        help="Frequency of the sine wave (default: %(default)s)",
    )
```
Then, we edit the `__init__` function too:
``` python
def __init__(self, args):
    self.times = np.linspace(0, 10, 100)
    self.values = np.sin(args.frequency * self.times)
```
Movie files have to have an `__init__` that takes two arguments. The second
is a `Namespace` that contains all the controllable options. These arguments
can be passed via command-line or configuration file.
``` sh
mopi -m sin_wave.py -o frames_dir --parallel --frequency 3
```
This command will produce the following video.

![sine_wave_fast](https://github.com/Sbozzolo/motionpicture/raw/master/sine_wave_fast.gif)

Alternatively, you can put any of arguments in a config file `conf`, for example:

``` text
outdir: frames_dir
frequency: 3
```
Config files support several syntaxes. Once you have the file, just call
``` sh
mopi sin_wave.py -c conf
```
You can use config files and command-line options at the same time, but in case
of conflict, the command-line arguments have the precedence.

## Unveiling data in an arbitrary file

Now that you have seen that you can control movies via command-line, it is time
to introduce you to the plugin system in `motionpicture`.

Suppose we have two-column files with time series data, we can modify the movie
file used in the previous example to animate those files, specifying which one
at run-time.

``` python
def mopi_add_custom_options(parser):
    """Add command-line options specific to this movie."""
    parser.add_argument(
        "-f",
        "--file",
        required=True,
        help="File to plot",
    )
```
Then, we `import numpy as np` and edit the `__init__` function too:
``` python
def __init__(self, args):
    self.times, self.values = np.loadtxt(args.file).T
    self.y_min, self.y_max = np.amin(self.value), np.amax(self.value)
```
We computed the minimum and maximum of the value so that we can adjust the y axis
range. The  `make_frame` method will be the same, with the exception that we change
the `plt.ylim([-1, 1])` line to `plt.ylim([self.y_min, self.y_max])`.

We can save this file as `plot_timeseries` and call `mopi`:
``` sh
mopi -m plot_timeseries -o frames -f my_file.dat
```
Of course, we can add as many options as we wish to control the output. For instance,
we may want to add a switch to use logarithmic axes instead. The class
`MOPIMovie` has full access to the user-supplied options, so you can do anything
you wish.

We did not hard-code anything in `plot_timeseries`, so the code will work for
any dataset. However, if we want to use this file again, but in a different
folder, we would have to copy it over, since `mopi -m` expects the path of the
movie file. Alternatively, we can copy `plot_timeseries` to a specific folder
of our choice, for example `~/.mopi_videos`. Then, we can set the environment
variable `MOPI_MOVIES_DIR` to be `~/.mopi_videos`, and `mopi` will be able to
find `plot_timeseries` from anywhere in your filesystem. In this case, you can
simply call:
``` sh
mopi plot_timeseries -o frames -f my_other_file.dat
```
Essentially, `plot_timeseries` became a plugin for `motionpicture` and you can
animate any data without having to write new code. This is one of the greatest
strengths of `motionpicture`, as it encourages you to write generic code that you
can easily reuse.

# Installation

`motionpicture` is available on PyPI. You can install it with `pip`:

``` sh
pip3 install motionpicture
```

To produce the final video, you have to have `ffmpeg` installed. Without
`ffmpeg`, you will not be able to glue together the frames, but you can still
use `motionpicture` to render the frames.

# Movie files

In the language of `motionpicture`, a movie file is a recipe on how to
generate an individual frame. It is completely up to you how you do that, but
`motionpicture` imposes some minimum requirements:

- It has to be a valid Python 3 file.
- It has to contain a class `MOPIMovie` with a method `make_frame` and a method
  `get_frames`.
- The method `__init__` has to take two arguments.
- The method `get_frames` has to return an iterable (e.g., a list) that
  identifies each frame. The elements of this iterable are passed as the `frame`
  argument to `make_frame`.
- The method `make_frame` has to take two arguments, the `path` of the output of
  the frame, and `frame`, the value that identifies frame (typically the frame
  number). `path` is where the image has to be saved. You are in charge of
  saving the image using the save method of your plotting package.

Other than these requirements, you can do anything you want in the movie file
(e.g., you can add more methods, functions, classes...).

To have support for the `--overwrite` option, the function `make_frame` must
always write the data regardless of possible pre-existing files at destination.

> :warning: Due to its own nature, `motionpicture` has to execute any code that
>           you supply. Do not use `motionpicture` with codes you do not trust!

# `mopi`

`mopi` is a command-line utility with several options. Its `--help` flag can
tell you what it can do:
``` text
General options:
  movie                 Movie to render among the ones found in MOPI_MOVIES_DIR. See bottom of the help message for list.
  -m MOVIE_FILE, --movie-file MOVIE_FILE
                        Path of the movie file.
  -c CONFIG, --config CONFIG
                        Config file path
  --movies-dir MOVIES_DIR
                        Folder where to look form movies.   [env var: MOPI_MOVIES_DIR]
  -o OUTDIR, --outdir OUTDIR
                        Output directory for frames and video.
  --snapshot SNAPSHOT   Only produce the specified snapshot (useful for testing).
  --overwrite           Overwrite files that already exist.
  --disable-progress-bar
                        Do not display the progress bar when generating frames.
  --parallel            Render frames in parallel.
  --skip-existing       Do not generate frames that already exist. No consistency checks are performed.
  --num-workers NUM_WORKERS
                        Number of cores to use (default: 8).
  --max-tasks-per-child MAX_TASKS_PER_CHILD
                        How many chunks does a worker have to process before it is respawned? Higher number typically leads to higher performance
                        and higher memory usage. (default: 1).
  --chunks-size CHUNKS_SIZE
                        How many frames does a worker have to do each time? Higher number typically leads to higher performance and higher memory usage.
  --only-render-movie   Do not generate frames but only render the final video.
  --frame-name-format FRAME_NAME_FORMAT
                        If only-render-movie is set, use this C-style frame name format instead of computing it. For example, '%04d.png' will
                        assemble a video with frames with names 0000.png, 0001.png, and so on, as found in the outdir folder.
  -v, --verbose         Enable verbose output.
  -h, --help            Show this help message and exit.

Frame selection:
  --min-frame MIN_FRAME
                        Do not render frames before this one.
  --max-frame MAX_FRAME
                        Do not render frames after this one.
  --frames-every FRAMES_EVERY
                        Render a frame every N (default: render all the possible frames).

Video rendering options:
  --movie-name MOVIE_NAME
                        Name of output video file, without extension (default: video).
  --extension EXTENSION
                        File extension of the video (default: mp4).
  --fps FPS             Frames-per-second of the video (default: 25).
  --codec CODEC         Codec to use for the final encoding. If not specified, it is determined from the file extension.
  --author AUTHOR       Author metadata in the final video.
  --title TITLE         Title metadata in the final video.
  --comment COMMENT     Comment metadata in the final video.

No movies found in the MOPI_MOVIES_DIR (.)
```

A useful option for debugging is `--snapshot`. If you pass the keyword
`--snapshot` and the identifier for a specific frame (an element of the iterable
`MOPIMoive.get_frames()`), `mopi` will only render that single frame. This
can be used to test your movie file.

Another interesting option is `--only-render-movie`. This skips the generation
of frames and only produces the final video. When this option is enable, `mopi`
will still go through the selection of frames from the movie file, so options
like `--min-frame` or `--frames-every` will affect the result. If you specify
also `--frame-name-format`, you can skip this step too (which skips the movie
file entirely), and just render the final video. This option requires a C-style
format string to specify which files have to be assembled to the final video.
This refers to the name of the files in the output folder.

# Development

We use:
* [Poetry](python-poetry.org) to manage dependencies, build, and publish
  `motionpicture`.
* [Black](https://github.com/psf/black) for formatting the code (with 89
  columns).
* [pytest](https://pytest.org) for unit tests (with `pytest-cov` for test
  coverage).
* GitHub actions for continuous integration.

We are happy to accept contributions.

## A note on the inner workings

Multiprocessing with Python is a pain. Hence, we need to hack our way lo support
parallelism in such a way that is completely transparent to the user. To achieve
this `motionpicture` uses two tricks:
- The movie file is evaluated verbatim with `exec` in the global namespace
- The `MOPIMovie` class is patched to a `pMOPIMovie` (always using `exec`) to
  provide a function `p_make_frames` that works better with `multiprocessing`.

So, `motionpicture` directly manipulates the global namespace to be able to
use multiple processes. This can lead to surprises in the code, for example,
`MOPIMovie` is used without apparently being imported.

# Changelog

See [NEWS.md](https://github.com/Sbozzolo/motionpicture/blob/master/NEWS.md) for
a changelog.

# Credits

The idea for `motionpicture` originated from the `SimVideo` package developed by
Wolfgang Kastaun.



NEWS.md
# Changelog

## Version 0.3.2 (13 April 2023)

Bug fixes:
- Improve compatibility with newer versions of Python

## Version 0.3.1 (6 July 2022)

Bug fixes:
- Check if movie files are Python files using the MIME type provided by libmagic

Other improvements:
- Recommend `--verbose` to see details of traceback

## Version 0.3.0 (29 April 2022)

New features added:
- `--codec`
- `--overwrite` (movie files have to implement this separately)
- `--skip-existing`

- `--maxtasksperchild`
- `--chunksize`

Other improvements:
- `mp4` files are better encoded and should be playable on more devices
- When `verbose`, now `ffmpeg` displays its messages
- Reworked parallelization scheme, now it is more efficient and less memory
  intensive.

Bug fixes:
- Do not require movie file when `--only-render-movie` and `--frame-name-format`
  are passed.

The minimum version of Python required is now Python 3.6.2.

## Version 0.2.0 (5 April 2021)

New features added:
- `--snapshot`
- `--frame-min`, `--frame-max`, `--frames-every`
- `--only-render-movie`, `--frame-name-format`

Other improvements:
- Now, `--verbose` prints more informative error messages, including the ones
  produced by `ffmpeg`

Bug fixes:
- The `--fps` flag now works as intended

## Version 0.1.2 (1 January 2021)

Initial release



tests/__init__.py
#!/usr/bin/env python3

# Copyright (C) 2020-2023 Gabriele Bozzola
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3 of the License, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, see <https://www.gnu.org/licenses/>.


tests/test_mopi.py
#!/usr/bin/env python3

# Copyright (C) 2020 - 2021 Gabriele Bozzola
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3 of the License, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, see <https://www.gnu.org/licenses/>.

from motionpicture import mopi


motionpicture/argparse.py
#!/usr/bin/env python3

# Copyright (C) 2020-2023 Gabriele Bozzola
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3 of the License, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, see <https://www.gnu.org/licenses/>.

"""In this module we have all the functions that have to do with dealing with
user-supplied arguments and finding the movie file.

The main function is :py:func:`~.get_args_movie`. This is function does all the
processing of the arguments, and imports the movie file. This function has the
non-trivial task of updating the argument list dynamically, depending on the
user-supplied options. For example, if a movie add custom command-line options, we must
display them in the help message.

A second function is :py:func:`~._init_argparse`. This function defines all the arguments
that can be customized via command-line. If you want to add new arguments, you have to
add them here. Notice, we do not use Python's argparse, but we use ``configargparse``.
This gives the user more flexibility on how to pass the arguments. Since there are tens
of options, users may want to have config files. Also, with ``configargparse`` some
options can be specified with a environment variable, which is nice for options like the
directory where to look for movies.

We also have :py:func:`~._is_movie_file`. This function parses a Python file and checks
if the requirements to be a "movie file" are met. These include containing a
``MOPIMovie`` class with specific methods. If new requirements arise, this function must
be edited.

The function :py:func:`~._check_args` checks the command-line arguments for possible
problems. It is wise to review this function if new options are added.

"""

import ast
import os
import sys
import warnings
from argparse import RawTextHelpFormatter

import configargparse
import magic


def _init_argparse(*args, **kwargs):
    """Initialize a new argparse parser and fill it with arguments.

    If you want to add new command-line options, here is where you have to look
    at.

    The arguments passed to this function and passed to the constructor of
    ``ArgParser``.

    All the arguments will be passed to the ``MOPIMovie`` class.

    :returns: Argparse parser with all the options except the custom ones added
              by the specific movie.
    :rtype: ``configargparse.ArgumentParser``

    """

    parser = configargparse.ArgParser(*args, **kwargs)

    general_options = parser.add_argument_group("General options")

    general_options.add(
        "movie",
        nargs="?",
        help="Movie to render among the ones found in MOPI_MOVIES_DIR. "
        "See bottom of the help message for list.",
    )
    general_options.add_argument("-m", "--movie-file", help="Path of the movie file.")
    general_options.add("-c", "--config", is_config_file=True, help="Config file path")
    general_options.add(
        "--movies-dir",
        default=".",
        help="Folder where to look form movies.",
        env_var="MOPI_MOVIES_DIR",
    )
    general_options.add_argument(
        "-o",
        "--outdir",
        default=".",
        help="Output directory for frames and video.",
    )
    general_options.add(
        "--snapshot",
        help="Only produce the specified snapshot (useful for testing).",
    )
    general_options.add(
        "--overwrite",
        action="store_true",
        help="Overwrite files that already exist.",
    )
    general_options.add(
        "--disable-progress-bar",
        action="store_true",
        help="Do not display the progress bar when generating frames.",
    )
    general_options.add(
        "--parallel", help="Render frames in parallel.", action="store_true"
    )
    general_options.add(
        "--skip-existing",
        help=(
            "Do not generate frames that already exist. "
            "No consistency checks are performed."
        ),
        action="store_true",
    )
    general_options.add(
        "--num-workers",
        default=os.cpu_count(),
        type=int,
        help="Number of cores to use (default: %(default)s).",
    )
    general_options.add(
        "--max-tasks-per-child",
        default=1,
        type=int,
        help="How many chunks does a worker have to process before it is"
        " respawned? Higher number typically leads to higher"
        " performance and higher memory usage. (default: %(default)s).",
    )
    general_options.add(
        "--chunk-size",
        default=1,
        type=int,
        help="How many frames does a worker have to do each time? Higher number"
        " typically leads to higher performance and higher memory usage.",
    )
    general_options.add(
        "--only-render-movie",
        help="Do not generate frames but only render the final video.",
        action="store_true",
    )
    general_options.add(
        "--frame-name-format",
        help="If only-render-movie is set, use this C-style frame name format"
        " instead of computing it. For example, '%%04d.png' will assemble a "
        "video with frames with names 0000.png, 0001.png, and so on, as found"
        " in the outdir folder.",
    )
    general_options.add(
        "-v", "--verbose", help="Enable verbose output.", action="store_true"
    )
    general_options.add_argument(
        "-h",
        "--help",
        action="store_true",
        help="Show this help message and exit.",
    )

    frames_options = parser.add_argument_group("Frame selection")

    frames_options.add_argument(
        "--min-frame",
        help="Do not render frames before this one.",
    )
    frames_options.add_argument(
        "--max-frame",
        help="Do not render frames after this one.",
    )
    frames_options.add_argument(
        "--frames-every",
        default="1",
        type=int,
        help="Render a frame every N (default: render all the possible frames).",
    )

    video_options = parser.add_argument_group("Video rendering options")

    video_options.add_argument(
        "--movie-name",
        default="video",
        help="Name of output video file, without extension (default: %(default)s).",
    )
    video_options.add_argument(
        "--extension",
        default="mp4",
        help="File extension of the video (default: %(default)s).",
    )
    video_options.add_argument(
        "--fps",
        default="25",
        type=int,
        help="Frames-per-second of the video (default: %(default)s).",
    )
    video_options.add_argument(
        "--codec",
        help="Codec to use for the final encoding."
        " If not specified, it is determined from the file extension.",
    )
    video_options.add_argument("--author", help="Author metadata in the final video.")
    video_options.add_argument("--title", help="Title metadata in the final video.")
    video_options.add_argument("--comment", help="Comment metadata in the final video.")
    # If you add new metadata, you have to update the function
    # _metadata_from_args in moviemaker.py

    return parser


def _is_movie_file(path):
    """Check if the file path fulfills the minimum requirement to be used by
    ``motionpicture``.

    These are:
    - It must contain a class ``MOPIMovie``,
    - The class ``MOPIMovie`` must have a method ``__init__`` which has to take
      exactly one argument.
    - The class ``MOPIMovie`` must have a method ``self.get_frames``,
    - The class ``MOPIMovie`` must have a method ``self.make_frame`` which has
      to take exactly two arguments

    To do this, we open the file and inspect the content with ``ast``.

    If new requirement arise, you have to add them here!

    :param path: Path of the file to inspect.
    :type path: str

    :returns: Return True if all the requirements are met, False if at least one
              is not. Along with this, return a string with the description of
              the requirement that was not met (if all are met, the string will be
              empty).
    :rtype: tuple with bool and string

    """
    # We assume that path has been checked somewhere else.
    type_ = magic.from_file(path)
    if not type_.startswith("Python"):
        return (
            False,
            f"File {path} is not a Python file, type: {type_}",
        )

    # First, we check if the file is Python file
    try:
        tree = ast.parse(open(path).read())
    except (SyntaxError, UnicodeDecodeError):
        return False, f"File {path} is not a valid Python file"

    # We check if there is a class ``MOPIMovie``.

    classes_in_tree = {node for node in tree.body if isinstance(node, ast.ClassDef)}

    movies = [class_ for class_ in classes_in_tree if class_.name == "MOPIMovie"]

    if len(movies) == 0:
        return False, f"File {path} does not contain a MOPIMovie class"

    # There must be one movie.
    movie_class = movies[0]

    methods_in_movie = {
        node for node in movie_class.body if isinstance(node, ast.FunctionDef)
    }

    init_method = [method for method in methods_in_movie if method.name == "__init__"]

    if len(init_method) == 0:
        return (
            False,
            f"MOPIMovie class in file {path} does not contain a method __init__",
        )

    num_args_init = len(init_method[0].args.args)
    if num_args_init != 2:
        return (
            False,
            f"__init__ in MOPIMovie class in file {path} takes {num_args_init} arguments, "
            "it should take 2 (self, args)",
        )

    get_frames_method = [
        method for method in methods_in_movie if method.name == "get_frames"
    ]

    if len(get_frames_method) == 0:
        return (
            False,
            f"MOPIMovie class in file {path} does not contain a method get_frames",
        )

    make_frame_method = [
        method for method in methods_in_movie if method.name == "make_frame"
    ]

    if len(make_frame_method) == 0:
        return (
            False,
            f"MOPIMovie class in file {path} does not contain a method make_frame",
        )

    # There must be one make_frame.
    # It has to have exactly two arguments
    num_args = len(make_frame_method[0].args.args)
    if num_args != 3:
        return (
            False,
            f"make_framke in MOPIMovie class in file {path} takes {num_args} arguments, "
            "it should take 3 (self, path, frame_number)",
        )

    return True, ""


def _search_for_movies(folder):
    """Search in the given ``folder`` for files that satisfy the requirement
    specified by the function :py:func:`~._is_movie_file`.

    :param folder: Folder to scan.
    :type folder: str

    :returns: Set of files that contain movies.
    :rtype: set
    """
    files_in_folder = {
        file_
        for file_ in os.listdir(folder)
        if os.path.isfile(os.path.join(folder, file_))
    }
    return {
        file_
        for file_ in files_in_folder
        if _is_movie_file(os.path.join(folder, file_))[0]
    }


def _check_args(args):
    """Check if there are problems with the provided arguments.

    We check:
    - If the number of workers is larger than the number of CPUs.

    :param args: Arguments to check.
    :returns: Arguments as read from command line or from args
    :rtype: ``argparse.Namespace``

    """

    if args.num_workers > os.cpu_count():
        warnings.warn(
            f"You requested {args.num_workers} cores, "
            f"but the machine only has {os.cpu_count()}. "
            "This may result in performance loss."
        )


def get_args_movie(namespace, cli_args=None):
    """Process arguments and import `MOPIMovie` into `namespace``.

    When executing this function, you are also going to execute the movie file.

    This function calls :py:func:`~._init_argparse` to setup the options
    and :py:func:`~._check_args` to check for problems.

    :param namespace: Namespace in which to execute the movie file.
    :type namespace: dict

    :param args: List of arguments as if they were passed via command-line.
                 This is used only for testing.
    :type args: list

    :returns: Arguments as read from command line or from ``args``.
    :rtype: ``argparse.Namespace``

    """

    if cli_args is None:
        # Remove the name of the program from the list of arguments
        cli_args = sys.argv[1:]

    # We do a two-step parsing of the command-line arguments. With the first step,
    # we:
    # 1. Understand if a movie was selected
    # 2. Check what movies are available in the MOPI_MOVIES_DIR
    # 3. Update the help message adding choices for the possible movies.
    #
    # With the second step we actually parse all the arguments.
    #
    # The detailed strategy is the following:
    # 1. We start with a parser that has all the arguments common to all the movies.
    #    In this, we remove the help flag, as we added our own.
    # 2. We parse the arguments. This is first pass is to understand which movie
    #    the user has selected.
    # 3. We load the movie module, and add the arguments for that movie.
    # 4. We parse again the arguments.

    desc = (
        "Make a video specifying all the details using command-line arguments. "
        "To use this utility, you have to specify a movie. "
        "This code will look for movies in the MOPI_MOVIES_DIR, which you can customize. "
        "To select one of these movies, just pass the file name as first argument. "
        "Alternatively, you can pass the argument -m and specify a file."
    )

    parser = _init_argparse(
        description=desc, add_help=False, formatter_class=RawTextHelpFormatter
    )

    # We have to use this function because the user may have passed arguments
    # that are don't know yet (since we haven't loaded the movie's argument yet)
    parsed, unknown = parser.parse_known_args(cli_args)

    # Now we try to load the movie, if possible. If the user specified a
    # movie-file, that has to have the precedence.

    if parsed.movie_file:
        movie_file = parsed.movie_file
    elif parsed.movie:
        movie_file = os.path.join(parsed.movies_dir, parsed.movie)
    else:
        movie_file = None

    # The case in which the user passed the arguments --only-render-movie and
    # --frame-name-format does not require a movie file. Which means that if the
    # user didn't pass --only-render-movie or --frame-name-format we need the
    # movie file.
    movie_file_required = (
        parsed.only_render_movie is False
    ) or parsed.frame_name_format is None

    if movie_file and movie_file_required:
        # Check if the specified movie file exists.
        if not (os.path.exists(movie_file) and os.path.isfile(movie_file)):
            raise ValueError(f"Movie-file {movie_file} does not exist.")

        # Check if the specified movie file meets the minimum requirements.
        is_movie_file, error_message = _is_movie_file(movie_file)
        if not is_movie_file:
            raise ValueError(error_message)

        # HACK: This is a nuclear way to achieve what we want.
        #
        # We want to import the user-supplied file. We could use importlib for
        # that. However, importing the movie as module introduces significant
        # complications in the parallelization. For instance, the imported file
        # is not in the PYTHONPATH of the various other sub-processes, which
        # makes impossible to pickle `MOPIMovie` (hence, it is impossible to use
        # multiprocessing!).
        #
        # The solution is to simply execute the file in the global scope. If we
        # do that, we can use multiprocessing without any problem.

        exec(open(movie_file).read(), namespace)

        # The user-supplied file can define a function `mopi_add_custom_options`
        # to add additional command-line options.
        if "mopi_add_custom_options" in namespace:
            custom_options = parser.add_argument_group("Movie custom options")
            namespace["mopi_add_custom_options"](custom_options)

    # Update the "help" adding the list of movies detected in the relevant folder.
    available_movies = _search_for_movies(parsed.movies_dir)

    if available_movies:
        epilog = f"Movies found in the MOPI_MOVIES_DIR ({parsed.movies_dir}):"
        for m in available_movies:
            epilog += f"\n* {m}"
    else:
        epilog = f"No movies found in the MOPI_MOVIES_DIR ({parsed.movies_dir})"

    parser.epilog = epilog

    if parsed.help:
        parser.print_help()
        sys.exit(0)

    if not movie_file and movie_file_required:
        raise ValueError("Movie not specified. Please, specify a movie.")

    args = parser.parse_args(cli_args)

    _check_args(args)

    return args
